lib/validate/validate_function.js

```diff
@@ -51,13 +53,29 @@ module.exports = function validateFunction(options) {
             return [new ValidationError(key, value, 'array length %d expected, length %d found', 2, value.length)];
         }
 
-        errors = errors.concat(validate({
-            key: key + '[0]',
-            value: value[0],
-            valueSpec: {type: 'number'},
-            style: options.style,
-            styleSpec: options.styleSpec
-        }));
+        var type = getType(value[0]);
+        if (!stopKeyType) stopKeyType = type;
+        if (type !== stopKeyType) {
+            return [new ValidationError(key, value, '%s stop key type must match previous stop key type %s', type, stopKeyType)];
+        }
+
+        if (type === 'object') {
+            errors = errors.concat(validate({
+                key: key + '[0].zoom',
+                value: value[0].zoom,
+                valueSpec: {type: 'number'},
+                style: options.style,
+                styleSpec: options.styleSpec
+            }));
```

Comment:
I think we can assert that

 - `value` is either a string or a number
 - no keys other than `value` and `zoom` are present